core/src/main/java/org/apache/iceberg/avro/TypeToSchema.java

```diff
@@ -86,9 +86,15 @@ public Schema struct(Types.StructType struct, List<Schema> fieldSchemas) {
     List<Schema.Field> fields = Lists.newArrayListWithExpectedSize(fieldSchemas.size());
     for (int i = 0; i < structFields.size(); i += 1) {
       Types.NestedField structField = structFields.get(i);
+      String origFieldName = structField.name();
+      boolean isValid = validAvroName(origFieldName);
+      String fieldName =  isValid ? origFieldName : sanitize(origFieldName);
       Schema.Field field = new Schema.Field(
-          structField.name(), fieldSchemas.get(i), null,
+          sanitize(fieldName), fieldSchemas.get(i), null,
```

Comment:
This calls sanitize twice if the name isn't valid.

Reply:
Fixed!.